Adding more test cases for testing multiple http methods

sebing opened the pull request:
Fixing the issue #458

Adding more tests for TkProxyTest

davvd commented:
@sebing I will find a reviewer for your pull requests shortly, thanks for contribution!

davvd commented:
@dmzaytsev review this one, please

dmzaytsev commented on src/test/java/org/takes/tk/TkProxyTest.java:
```diff
@@ -34,34 +34,69 @@
     /**
-     * TkProxy can work.
-     * @throws Exception If some problem inside
+     * TkProxy Testcase for testing the creation of the object with method POST.
```
@sebing please update java doc according to this article www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html?expand=1#test-method-names

same below

dmzaytsev commented:
@sebing see my comment above

sebing commented:
@dmzaytsev  thanks for comments. I corrected them. 
And thanks for the link, it had more guidelines(that clear lot of doubts). I like them and follow them.

dmzaytsev commented:
@sebing thank you

dmzaytsev commented:
@rultor merge

rultor commented:
> @rultor merge

@dmzaytsev Thanks for your request. @yegor256 Please confirm this.

yegor256 commented on src/test/java/org/takes/tk/TkProxyTest.java:
```diff
@@ -34,34 +34,69 @@
+     * @param httpmethod HTTP methods (GET, POST, etc),
+     * @throws Exception If some problem inside.
+     */
+    private void justWorks(final String httpmethod) throws Exception {
```
@sebing why not just `method`? see http://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html

yegor256 commented on src/test/java/org/takes/tk/TkProxyTest.java:
```diff
@@ -71,11 +106,13 @@ public void exec(final URI home) throws IOException {
      */
-    @Test
-    public void correctlyMapsPathString() throws Exception {
+    private void correctlyMapsPathString(final String httpmethod)
```
@sebing same here, what's wrong about `method`?

yegor256 commented:
@sebing a few comments, see above

sebing commented:
@yegor256 Fixed the stingy compound names.

sebing commented:
@yegor256 @davvd Any update on the closure of this merge?

yegor256 commented:
@sebing I don't like this design of test methods. I should probably write an article about it, but here is what I mean, in short. This is your code:

```
@Test
void doesThis() {
  this.works(1);
}
@Test
void doesThat() {
  this.works(2);
}
```

This is how it should be instead:

```
@Test
void doesThis() {
  assertThat(this.asset(1), is(true));
}
@Test
void doesThat() {
  assertThat(this.asset(2), is(true));
}
```

My point is that you should **not** create a test method that doesn't have an assertion. By doing that you completely destroys the idea of a unit test. The test becomes unclear and unmaintainable.

sebing commented:
@yegor256 I agree with you.
But I thought this is doing the same.

```java
MatcherAssert.assertThat(
                        new RsPrint(
                            new TkProxy(home).act(new RqFake(method))
                        ).print(),
                        Matchers.containsString("hello")
```
@dmzaytsev
To make the code more redable,
this assertion is happening in a private method, should i move it to the test case? what do you think of this?

sebing commented:
@dmzaytsev  what do you think? How can we improve the code?

sebing commented:
@yegor256 @dmzaytsev  I know, this is not a school :) still...
- What is the difference between assertThat in the test  and in a private method?
- Is it bad to assert in a private method? ( i cannot find a clear best practice in google)

sebing commented:
@yegor256 If the code with assert in the private methods, can you accept the code?

dmzaytsev commented:
@sebing just do it :)

https://github.com/yegor256/takes/pull/551#issuecomment-178987626

> yegor256 commented 2 days ago
> My point is that you should not create a test method that doesn't have an assertion. By doing that you completely destroys the idea of a unit test. The test becomes unclear and unmaintainable.

sebing commented:
@dmzaytsev Can you re-review the changes?

dmzaytsev commented:
@sebing If I had to do that task, I would suggest something like this
```
final String [] methods = {RqMethod.POST, RqMethod.GET, .....};
final String content = "content";
for(final String method: methods) {
    new FtRemote(new TkFixed(content)).exec(
            new FtRemote.Script() {
                @Override
                public void exec(final URI home) throws IOException {
                    MatcherAssert.assertThat(
                       String.format("method %s", method),
                        new RsPrint(
                            new TkProxy(home).act(new RqFake(method))
                        ).print(),
                        Matchers.containsString(content)
                    );
                }
            }
        );
}
```

sebing commented:
@dmzaytsev Is there any other way we can test?
I am getting this AppVeyor check error:
```
Avoid instantiating new objects inside loops
```

dmzaytsev commented:
@sebing `@SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")`

sebing commented:
@dmzaytsev It is a better implementation with the loop. 
May be i was deceived by the task name(but no excuse). Can you review the test again?

dmzaytsev commented on src/test/java/org/takes/tk/TkProxyTest.java:
```diff
@@ -50,30 +51,43 @@
-     * TkProxy can work.
+     * An array of http methods for testing.
+     */
+    private static final String[] METHODS = {
```
@sebing what about other http methods ?

sebing replied:
@dmzaytsev The requirement was not clear for me :) 
I added the methods except( CONNECT and HEAD, they need a different testing).
Should I add the testcase for them too?

sebing replied:
@dmzaytsev  I will add another test case for HEAD
And regarding the Connect
```
java.io.IOException: Failed CONNECT request to http://localhost:52991/

	at com.jcabi.http.request.JdkRequest$1.send(JdkRequest.java:127)
	at com.jcabi.http.request.BaseRequest.fetchResponse(BaseRequest.java:336)
	at com.jcabi.http.request.BaseRequest.fetch(BaseRequest.java:251)
	at org.takes.tk.TkProxy.act(TkProxy.java:119)
	at org.takes.tk.TkProxyTest$2.exec(TkProxyTest.java:94)
	at org.takes.http.FtRemote.exec(FtRemote.java:157)
	at org.takes.tk.TkProxyTest.justWorksOnConnect(TkProxyTest.java:90)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
Caused by: java.net.ProtocolException: Invalid HTTP method: CONNECT
	at java.net.HttpURLConnection.setRequestMethod(HttpURLConnection.java:428)
	at com.jcabi.http.request.JdkRequest$1.send(JdkRequest.java:102)
	... 32 more
```
Should I add a puzzle?

dmzaytsev commented:
@sebing just one comment above

dmzaytsev commented on src/test/java/org/takes/tk/TkProxyTest.java:
```diff
@@ -46,24 +47,70 @@
  * @todo #377:30min/DEV We need more tests for TkProxy.
```
@sebing you have to delete the original puzzle

dmzaytsev commented on src/test/java/org/takes/tk/TkProxyTest.java:
```diff
@@ -46,24 +47,70 @@
  *  The tests should verify the different HTTP methods (GET, POST, etc),
  *  as well as the different combinations of request/response headers.
+ *
+ * @todo #458:30min We need to support the http method CONNECT.
```
@sebing I think we don't need to proxy `CONNECT` requests

dmzaytsev commented on src/test/java/org/takes/tk/TkProxyTest.java:
```diff
@@ -46,24 +47,70 @@
+    /**
+     * The expected result for the Head http request.
+     */
+    private static final String RESULT = "";
```
@sebing why you need this variable?

dmzaytsev commented on src/test/java/org/takes/tk/TkProxyTest.java:
```diff
@@ -46,24 +47,70 @@
+                            )
                         ).print(),
-                        Matchers.containsString("hello")
+                        Matchers.containsString(TkProxyTest.RESULT)
```
@sebing I didn't get what you check here ?
I guess any string contains `""`, isn't it?

Anyway I think the existing methods enough for the test, let's remove test for  `HEAD`

dmzaytsev commented:
@sebing please see my comments

sebing commented:
@dmzaytsev I corrected the suggestions. 
- Removed the HEAD and Connect methods.
- Removed the puzzle.
You have any idea one of the travis-ci(java 8) job is failing?

dmzaytsev commented:
@sebing try close/open this PR to trigger a new Travis build

sebing commented:
@dmzaytsev trying to re run the travis

dmzaytsev commented:
@sebing feel free to submit a bug about the unstable test on java 8

sebing commented:
@dmzaytsev Now the build is success. Can you review them?

dmzaytsev commented:
@sebing looks good
thanks

dmzaytsev commented:
@rultor merge

rultor commented:
> @rultor merge

@dmzaytsev Thanks for your request. @yegor256 Please confirm this.

sebing commented:
@yegor256 Any update on this PR?

sebing commented:
@yegor256 Can you look at this PR?

yegor256 commented:
@rultor try to merge

rultor commented:
> @rultor try to merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/6645-181697755)

rultor commented:
> @rultor try to merge

@sebing @yegor256 Oops, I failed. You can see the full log [here](http://www.rultor.com/t/6645-181697755) (spent 16min)

```
INFO: "reading src/test/java/org/takes/rs/RsWithCookieTest.java..."
INFO: "reading src/test/java/org/takes/rs/RsWithHeaderTest.java..."
INFO: "reading src/test/java/org/takes/rs/RsWithHeadersTest.java..."
INFO: "reading src/test/java/org/takes/rs/RsWithStatusTest.java..."
INFO: "reading src/test/java/org/takes/rs/RsWithTypeTest.java..."
INFO: "reading src/test/java/org/takes/rs/RsWithoutHeaderTest.java..."
INFO: "reading src/test/java/org/takes/rs/RsXSLTTest.java..."
INFO: "reading src/test/java/org/takes/rs/package-info.java..."
INFO: "reading src/test/java/org/takes/rs/xe/RsXemblyTest.java..."
INFO: "reading src/test/java/org/takes/rs/xe/XeAppendTest.java..."
INFO: "reading src/test/java/org/takes/rs/xe/XeDateTest.java..."
INFO: "reading src/test/java/org/takes/rs/xe/XeLinkHomeTest.java..."
INFO: "reading src/test/java/org/takes/rs/xe/XeLinkSelfTest.java..."
INFO: "reading src/test/java/org/takes/rs/xe/XeLinkTest.java..."
INFO: "reading src/test/java/org/takes/rs/xe/XeLocalhostTest.java..."
INFO: "reading src/test/java/org/takes/rs/xe/XeSLATest.java..."
INFO: "reading src/test/java/org/takes/rs/xe/XeTransformTest.java..."
INFO: "reading src/test/java/org/takes/rs/xe/XeWhenTest.java..."
INFO: "reading src/test/java/org/takes/rs/xe/package-info.java..."
INFO: "reading src/test/java/org/takes/tk/TkCORSTest.java..."
INFO: "reading src/test/java/org/takes/tk/TkClasspathTest.java..."
INFO: "reading src/test/java/org/takes/tk/TkFilesTest.java..."
INFO: "reading src/test/java/org/takes/tk/TkGzipTest.java..."
INFO: "reading src/test/java/org/takes/tk/TkHTMLTest.java..."
INFO: "reading src/test/java/org/takes/tk/TkMeasuredTest.java..."
INFO: "reading src/test/java/org/takes/tk/TkProxyTest.java..."
INFO: "reading src/test/java/org/takes/tk/TkReadAlwaysTest.java..."
INFO: "reading src/test/java/org/takes/tk/TkRedirectTest.java..."
INFO: "reading src/test/java/org/takes/tk/TkRetryTest.java..."
INFO: "reading src/test/java/org/takes/tk/TkSlf4jTest.java..."
INFO: "reading src/test/java/org/takes/tk/TkTextTest.java..."
INFO: "reading src/test/java/org/takes/tk/TkVerboseTest.java..."
INFO: "reading src/test/java/org/takes/tk/TkVersionedTest.java..."
INFO: "reading src/test/java/org/takes/tk/TkWithHeadersTest.java..."
INFO: "reading src/test/java/org/takes/tk/package-info.java..."
INFO: "reading src/test/resources/log4j.properties..."
INFO: "reading src/test/resources/org/takes/rs/inclass/imported.xsl..."
INFO: "reading src/test/resources/org/takes/rs/included.xsl..."
INFO: "reading src/test/resources/org/takes/rs/includes.xsl..."
INFO: "reading src/test/resources/org/takes/rs/simple.xsl..."
INFO: "reading src/www.takes.org/index.html..."
+ mv /home/r/repo .
++ whoami
+ chown -R root repo
+ '[' -n '' ']'
++ whoami
+ sudo chown -R rultor repo
+ cd repo
+ git push origin master
To git@github.com:yegor256/takes.git
 ! [rejected]        master -> master (fetch first)
error: failed to push some refs to 'git@github.com:yegor256/takes.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first merge the remote changes (e.g.,
hint: 'git pull') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
container 39d0d706a0687d5f0e0d7c18f848f7ed05e3e6036a98a0f38350a1c03dade968 is dead
Tue Feb  9 04:17:13 UTC 2016

```

sebing commented:
@yegor256  can we try to merge this again?

sebing commented:
@yegor256 The merge failed, can we try to merge this again?

sebing commented:
@yegor256 @davvd The rultor merge failed, can we try again?

sebing commented:
@yegor256 @dmzaytsev @davvd What can we do here? should we try to ask rultor to merge again?

dmzaytsev commented:
@sebing only @yegor256 can confirm the merge request as hi is the architect on this project
we should wait until Yegor return from a business trip

sebing commented:
@dmzaytsev Thanks for the information. 
@davvd is the information that the architect is not available published anywhere? So I won't be keep spamming people without this information next time? :)

yegor256 commented:
@rultor try to merge again

rultor commented:
> @rultor try to merge again

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/6645-184799518)

rultor commented:
> @rultor try to merge again

@yegor256 Done! FYI, the full log is [here](http://www.rultor.com/t/6645-184799518) (took me 35min)

davvd commented:
@elenavolokhova please, help us to evaluate quality here, as per [par.24](http://at.teamed.io/policy.html#24)

davvd commented:
@rultor deploy pls

rultor commented:
> @rultor deploy pls

@davvd OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/6645-185839123)

rultor commented:
> @rultor deploy pls

@davvd Done! FYI, the full log is [here](http://www.rultor.com/t/6645-185839123) (took me 9min)

elenavolokhova commented:
@davvd Looks good!

davvd commented:
> @davvd Looks good!

@elenavolokhova thanks for the review

davvd commented:
@dmzaytsev I added 10 mins to @elenavolokhova (for QA review) in transaction 78630815; thanks, I just added **34 mins** to your account, payment `78630822`, 460 hours and 9 mins spent; review comments (c=19) added as a bonus; +34 added to your rating, at the moment it is: [+3514](http://www.netbout.com/b/35166?open=rating)

sebing commented:
@dmzaytsev Can you help me understand this. This PR is merged and deployed. But the puzzle is still in the master. Any clue why?

dmzaytsev commented:
@sebing no idea
but I would suggest you create a new PR and remove the puzzle

sebing commented:
@dmzaytsev it is not the puzzle alone, the whole changes are not merged. So should I redo the work?

dmzaytsev commented:
@yegor256 take a look please

original-brownbear commented:
@sebing @dmzaytsev this is not Rultor. Please look at 79be7637a3e6eed12c0f666988c0fc8151e20482
The last commit to t his branch was simply an apparently wrong merge with upstream reverting a bunch of the prior changes.
I mean Rultor is not setting the "merged" status anyhow, this is GitHub. Unless you see a commit by Rultor, this cannot be Rultor's fault.

Makes sense?

sebing commented:
@dmzaytsev @original-brownbear So how do we proceed now?

original-brownbear commented:
@sebing you need to create a new PR and ask for another review. Obviously this one was wrong and got merged accidentally.
I suggest simply going to the commit before your 79be763 and correctly rebasing that onto `master`.
For all other questions on this your architect here, @yegor256 is who you should though ;)

yegor256 commented:
@sebing looks like it was your mistake, as an author of PR. just do it again.
